File: Examples/statistics/codes/fit.cpp
// ========================================================================================
// Example code: how to perform a Bayesian fit to a set of data points with a generic model
// ========================================================================================

#include "Cosmology.h"
#include "Data1D.h"
#include "Posterior.h"

using namespace std;

// these two variables contain the name of the CosmoBolognaLib
// directory and the name of the current directory (useful when
// launching the code on remote systems)
string cbl::par::DirCosmo = DIRCOSMO, cbl::par::DirLoc = DIRL;


// =====================================================================


// this example model has 4 parameters: A, B, C, D; A and B are free parameters, C is fixed, D is a derived parameter

vector<double> model_function (const vector<double> x, const shared_ptr<void> modelInput, std::vector<double> &parameter)
{
  // the object Cosmology, used in this example to compute Omega_matter
  cbl::cosmology::Cosmology cosm = *static_pointer_cast<cbl::cosmology::Cosmology>(modelInput);

  vector<double> model(x.size(), 0.);
  for (size_t i=0; i<x.size(); ++i)
    model[i] = parameter[0]*x[i]+parameter[1]+parameter[2]*cosm.Omega_matter(); // the model

  parameter[3] = parameter[0]+parameter[1]+parameter[2]; // parameter[3] is a derived parameter

  return model;
}


// =====================================================================


int main () {

  try {

    // --- set the input/output file/directories ---
    
    const string dir_input = cbl::par::DirLoc+"../input/";
    const string dir_output = cbl::par::DirLoc+"../output/";
    const string file_data = "data.dat";
    const string file_output_start = "model_starting_values.dat";
    const string file_output_bestfit = "model_bestfit.dat";
    

    // --- construct the dataset by reading an input file ---
    
    const cbl::data::Data1D data(dir_input+file_data);
    shared_ptr<cbl::data::Data> ptr_data = make_shared<cbl::data::Data1D>(data);

    
    // --- set the model to construct the likelihood ---
    
    // number of model parameters
    const int nparameters = 4;

    // names of the model parameters
    const vector<string> parNames = {"A", "B", "C", "D"};
    
    // starting values for A and B, and fixed values for C
    double valA = 1., valB = 1., valC = 1.; 
     
    // vector containing the 4 model parameters
    vector<cbl::statistics::ParameterType> parType(nparameters-1, cbl::statistics::ParameterType::_Base_);
    parType.emplace_back(cbl::statistics::ParameterType::_Derived_);   

    // set the stuff used to construct the model: here an object of class cosmology, just as an example 
    const cbl::cosmology::Cosmology cosmology;
    auto ptr_modelInput = make_shared<cbl::cosmology::Cosmology>(cosmology);

    // construct the model
    const cbl::statistics::Model1D model(&model_function, nparameters, parType, parNames, ptr_modelInput);
    auto ptr_model = make_shared<cbl::statistics::Model1D>(model);


    // --- construct and maximize the likelihood ---
    
    // define a Gaussian likelihood
    cbl::statistics::Likelihood likelihood(ptr_data, ptr_model, cbl::statistics::LikelihoodType::_Gaussian_Error_);

    // write the model used to construct the likelihood, at the starting parameter values (A=1, B=1, C=1=fixed)
    const vector<double> start = {valA, valB};
    likelihood.parameters()->fix(2, valC); // fix the parameter C, i.e. parameter[2], to valC=1
    likelihood.write_model(dir_output, file_output_start, start);

    // limits for A and B
    const double minA = -10., maxA = 10.;
    const double minB = -10., maxB = 10.;
    const vector<vector<double>> limits = { {minA, maxA} , {minB, maxB} };
    
    // maximize the likelihood and write the output
    likelihood.maximize(start, limits);
    likelihood.write_model_at_bestfit(dir_output, file_output_bestfit);

    
    // --- construct the priors ---
    
    // (remember to give different seed to priors!)
    auto prior_A = make_shared<cbl::statistics::PriorDistribution>(cbl::statistics::PriorDistribution(cbl::glob::DistributionType::_Uniform_, minA, maxA, 666));
    auto prior_B = make_shared<cbl::statistics::PriorDistribution>(cbl::statistics::PriorDistribution(cbl::glob::DistributionType::_Uniform_, minB, maxB, 999));
    auto prior_C = make_shared<cbl::statistics::PriorDistribution>(cbl::statistics::PriorDistribution(cbl::glob::DistributionType::_Constant_, valC));
    const vector<shared_ptr<cbl::statistics::PriorDistribution>> prior_distributions = {prior_A, prior_B, prior_C};

    
    // --- construct, maximize and sample the posterior ---
    
    cbl::statistics::Posterior posterior(prior_distributions, likelihood, 696);

    // maximize the posterior
    //posterior.maximize({valA, valB});

    // sample the posterior (starting the MCMC chain from the maximum of the posterior to speed up the chain convergence)
    const int nwalkers = 10;
    const int chain_size = 5000;
    posterior.initialize_chains(chain_size, nwalkers, 1.e-5, {valA, valB});
    posterior.sample_stretch_move(2);

    // show the median MCMC values of the four parameters on screen
    cout << endl;
    for (size_t i=0; i<posterior.parameters()->nparameters(); ++i)
      cout << setprecision(4) << "Posterior median of " << posterior.parameters()->name(i) << " = " << posterior.parameters()->bestfit_value(i) << endl;

    // show all the MCMC statistics on screen
    const int burn_in = 0;
    const int thin = 1;
    posterior.show_results(burn_in, thin);

    // store the chain ouputs
    posterior.write_results(cbl::par::DirLoc+"../output/", "chains_linear_relation", burn_in, thin);

    // store the best-fit model
    posterior.write_model_from_chain(cbl::par::DirLoc+"../output/", "model_from_chain.dat", {}, {}, burn_in, thin);

  }
  
  catch(cbl::glob::Exception &exc) { cerr << exc.what() << endl; exit(1); }

  return 0;
}

File: Examples/statistics/codes/fit.py
# ===================================================================
# Example code: how to perform a Bayesian fit to a set of data points
# with a generic model; to run this python script, first compile the
# wrapper with: make modelpy
# ===================================================================

# to ensure compatibility in Python versions 2.x and 3.x
from __future__ import print_function

# import Python modules for scientific computing 
import os
import numpy as np
import pandas as pd
# import Python modules for plotting
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

# import the CosmoBolognaLib modules 
import CosmoBolognaLib as cbl 

# import the wrapper that returns an object of type Model1D
from modelpy import getModel1D

try:
    import corner
except ImportError:
    print("This script requires the corner.py package.")
    print("Please install it if you are interested in producing corner plots")
    print("https://corner.readthedocs.io/en/latest/pages/custom.html")
    pass



# set the input/output file/directories 
dir_input = "../input/"
dir_output = "../output/"
file_data1 = "data1.dat"
file_data2 = "data2.dat"



def go (data, model, nwalkers, chain_size, results_name, model_name):

    # define a Gaussian likelihood
    likelihood = cbl.Likelihood(data, model, cbl.LikelihoodType__Gaussian_Error_)

    # set the model parameter
    valA, valB, valC = 1., 2., 1. 

    # limits for A and B
    minA, maxA = -10., 10.
    minB, maxB = -10., 10.

    # construct the priors 
    prior_A = cbl.PriorDistribution(cbl.DistributionType__Uniform_, minA, maxA)
    prior_B = cbl.PriorDistribution(cbl.DistributionType__Uniform_, minB, maxB)
    prior_C = cbl.PriorDistribution(cbl.DistributionType__Constant_, valC)
    prior_distributions = cbl.PriorDistributionPtrVector([prior_A, prior_B, prior_C])  

    # construct the posterior
    posterior = cbl.Posterior(prior_distributions, likelihood, 696)

    # sample the posterior (starting the MCMC chain from the maximum of the posterior to speed up the chain convergence)
    posterior.initialize_chains(chain_size, nwalkers, 1.e-5, [valA, valB])
    posterior.sample_stretch_move(2)

    # show the median MCMC values of the four parameters on screen
    print ("\n")
    for i in range(posterior.parameters().nparameters()):
        print("Posterior median of %s = %g\n"%(posterior.parameters().name(i), posterior.parameters().bestfit_value(i)))
    
    # store the chain ouputs
    posterior.write_results("../output/", results_name)
    
    return posterior



def plot_contours (posterior, burn_in, thin, figure, axes, color, label):

    # read the chains
    nPar = posterior.parameters().nparameters()
    table = np.array( [posterior.parameters().parameter_chain_values(i, burn_in, thin) for i in range(nPar)] )
    names = [posterior.parameters().name(i) for i in range(nPar)]
    status = [posterior.parameters().status(i) for i in range(nPar)]
    weights = posterior.weight(burn_in, thin)

    # select only non-fixed parameters
    from itertools import compress
    notFixed = list(map(lambda s : s!="FIXED", status))
    notFixedNames = list(compress(names, notFixed))
    notFixedChains = np.array(list(compress(table, notFixed))).T

    corner.corner(notFixedChains, \
                  weights=np.array(weights),\
                  labels=notFixedNames,\
                  truths=(1, 2., 4.), truth_color='black', \
                  plot_datapoints=False, show_titles=True, smooth=True, plot_density=False, levels=(1-np.exp(-0.5), 1-np.exp(-2), 1-np.exp(-4.5)),\
                  color=color, fill_contours=True, hist_kwargs={"density" : True, "color":color}, fig=figure)

    axes[0][1].plot(1,1,color, label=label)
    axes[0][1].legend(loc="best", fontsize=11)



# function that joines two chains: one after the other
def join_probes(posterior,dir_input,file_data_1, file_data_2, nwalkers):

    file_name_1 = dir_input+file_data_1
    file_name_2 = dir_input+file_data_2
    imp_chain_1 = np.loadtxt(file_name_1)
    imp_chain_2 = np.loadtxt(file_name_2)
    samples_1 = pd.DataFrame(imp_chain_1[:,:])
    samples_2 = pd.DataFrame(imp_chain_2[:,:])
    joined_samples = pd.concat([samples_1,samples_2])
    joined_file_name = "joined_impsamp.dat"
    open(dir_input+joined_file_name,"w").close()
    joined_samples.to_csv(dir_input+joined_file_name,header=None,index=None,sep='\t',mode="w")

    joined_posterior = posterior
    joined_posterior.read_chain(dir_input, joined_file_name,nwalkers)

    return joined_posterior



# set parameters for posterior sampling
nwalkers = 100
chain_size = 500
burn_in = 0
thin = 1

# set the stuff used to construct the model: here an object of class cosmology, just as an example 
cosmology = cbl.Cosmology()

# construct the dataset by reading an input file 
data1 = cbl.Data1D(dir_input+file_data1)

# set the model to construct the likelihood
model1 = getModel1D(cosmology, 0)
posterior1 = go(data1, model1, nwalkers, chain_size, "model1", "model1.dat")

# construct the dataset by reading an input file 
data2 = cbl.Data1D(dir_input+file_data2)

# set the model to construct the likelihood
model2 = getModel1D(cosmology, 1)
posterior2 = go(data2, model2, nwalkers, chain_size, "model2", "model2.dat")

# do the importance sampling and plot the contours

fig, axes = plt.subplots(3, 3, figsize=(15, 15))

plot_contours(posterior1, burn_in, thin, fig, axes, "b", "Posterior 1")
plot_contours(posterior2, burn_in, thin, fig, axes, "y", "Posterior 2")

# do the importance sampling
posterior1.importance_sampling("../output/", "model2_chain.dat")

# store the chain ouputs
posterior1.write_results("../output/", "model_1+2_importance_sampling")

# plot the contours
plot_contours(posterior1, burn_in, thin, fig, axes, "g", "imp. sampling 1+2")

# do the importance sampling
posterior2.importance_sampling("../output/", "model1_chain.dat")

# store the chain outputs
posterior2.write_results("../output/", "model_2+1_importance_sampling")

# plot the contours
plot_contours(posterior2, burn_in, thin, fig, axes, "r", "imp. sampling 2+1")

file_impsamp1 = "model_1+2_importance_sampling_chain.dat"
file_impsamp2 = "model_2+1_importance_sampling_chain.dat"

# join chains from importance sampling
joined_posterior = join_probes(posterior1, dir_output, file_impsamp1, file_impsamp2, 1)

# plot the contours
plot_contours(joined_posterior,burn_in, thin, fig, axes, "k", "imp. sampling concat.")

plt.show()
